table/src/main/java/org/enso/table/operations/OrderBuilder.java

```diff
@@ -0,0 +1,99 @@
+package org.enso.table.operations;
+
+import org.enso.table.data.column.storage.Storage;
+import org.enso.table.data.mask.OrderMask;
+import org.enso.table.data.table.Column;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.IntStream;
+
+/** Builds an order mask resulting in sorting storages according to specified rules. */
+public class OrderBuilder {
+  public static class OrderRule {
+    private final Column column;
+    private final Comparator<Object> customComparator;
+    private final boolean ascending;
+    private final boolean missingLast;
+
+    /**
+     * A single-column ordering rule.
+     *
+     * @param column the column to use for ordering
+     * @param customComparator a comparator that should be used instead of natural ordering of the
+     *     values
+     * @param ascending whether column should be sorted ascending or descending
+     * @param missingLast whether or not missing values should be placed at the start or end of the
+     *     ordering
+     */
+    public OrderRule(
+        Column column,
+        Comparator<Object> customComparator,
+        boolean ascending,
+        boolean missingLast) {
+      this.column = column;
+      this.customComparator = customComparator;
+      this.ascending = ascending;
+      this.missingLast = missingLast;
+    }
+
+    /**
+     * Builds an index-comparing comparator, that will sort array indexes according to the specified
+     * ordering of the underlying column.
+     *
+     * @param fallbackComparator a base value comparator, used in case the column does not define a
+     *     natural ordering.
+     * @return a comparator with properties described above
+     */
+    public Comparator<Integer> toComparator(Comparator<Object> fallbackComparator) {
+      final Storage storage = column.getStorage();
+      Comparator<Object> itemCmp = customComparator;
+      if (itemCmp == null) {
+        itemCmp = column.getStorage().getDefaultComparator();
+      }
+      if (itemCmp == null) {
+        itemCmp = fallbackComparator;
+      }
+      if (!ascending) {
+        itemCmp = itemCmp.reversed();
+      }
+      if (missingLast) {
+        itemCmp = Comparator.nullsLast(itemCmp);
+      } else {
+        itemCmp = Comparator.nullsFirst(itemCmp);
+      }
+
+      final Comparator<Object> cmp = itemCmp;
+      Comparator<Integer> result =
+          (i, j) -> cmp.compare(storage.getItemBoxed(i), storage.getItemBoxed(j));
```

Comment:
Couldn't we somehow avoid boxing here?

It seems that performance when sorting on primitive values (int, bool) will likely be heavily degraded by boxing here and within the IntStream. Are we ok with that? Is there a plan to improve it in the future?

Reply:
No, java does not support sorting primitive arrays with custom comparators. Also every comparator call already does a memory jump, so I don't think boxing hurts as at all.

Reply:
I see.

Still it could handle the special case when we sort in natural order, but hmm, probably as we'll be moving around the rows a lot it may not be such a big difference anyway. Maybe it could be filed as a possible low priority enhancement?

Reply:
yeah it could be a future enahncement